### PR TITLE
Mes 4037/mes 4268 debrief2 and vehicle checks for cat c

### PR DESCRIPTION
**Pull request by amobarak**
## Description

Debrief page 2 for cat C, vehicle checks logic for cat C.

(Can't reach debrief page until TODO MES-4287 is done, tested manually using BE workflow then going to cat C debrief)

## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [ ] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/modules/tests/pass-completion/cat-c/__tests__/pass-completion.reducer.cat-c.spec.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    expect(result.provisionalLicenceProvided).toBe(true);
+    result = passCompletionCatCReducer(result, new ProvisionalLicenseNotReceived());
+    expect(result.provisionalLicenceProvided).toBe(false);
+  });
+
+  it('should put that the provisional license was not received into state when no selected', () => {
+    let result;
+    result = passCompletionCatCReducer(result, new ProvisionalLicenseNotReceived());
+    expect(result.provisionalLicenceProvided).toBe(false);
+  });
```
Comment (rossfauldsbjss) [inline src/modules/tests/pass-completion/cat-c/__tests__/pass-completion.reducer.cat-c.spec.ts]: duplicated test not seeing difference to test on line 26

File: src/providers/fault-count/__tests__/fault-count.spec.ts [hunk 2]
```diff
@@ -139,4 +161,55 @@ describe('FaultCountProvider', () => {
...
+            description: 'string',
+            outcome: 'S',
+          }],
+        },
+        eyesightTest: {
+          complete: true,
+          seriousFault: true,
+        },
+      };
+      expect((faultCountProvider as any).getSeriousFaultSumCountCatBE(failedState)).toBe(2);
```
Comment (rossfauldsbjss) [inline src/providers/fault-count/__tests__/fault-count.spec.ts:188]: showMeQuestions shouldn't have an S - they are all driving faults, and the logic for if a serious fault is calculated is determined elsewhere.

Comment (rossfauldsbjss) [reply src/providers/fault-count/__tests__/fault-count.spec.ts:188]: So this test doesn't reflect a situation that would occur in the app.

File: src/providers/fault-count/__tests__/fault-count.spec.ts [hunk 3]
```diff
@@ -139,4 +161,55 @@ describe('FaultCountProvider', () => {
...
+            outcome: 'DF',
+          }],
+          showMeQuestions: [{
+            code: 'string',
+            description: 'string',
+            outcome: 'D',
+          }],
+        },
+      };
+      expect((faultCountProvider as any).getDangerousFaultSumCountCatBE(failedState)).toBe(1);
```
Comment (rossfauldsbjss) [inline src/providers/fault-count/__tests__/fault-count.spec.ts:212]: Similar to comment about serious fault (see above)

File: src/providers/fault-count/fault-count.ts [hunk 4]
```diff
@@ -191,6 +242,24 @@ export class FaultCountProvider {
...
 
+  private getSeriousFaultSumCountCatC = (data: CatCUniqueTypes.TestData): number => {
+
+    // The way how we store serious faults differs for certain competencies
+    // Because of this we need to pay extra attention on summing up all of them
+    const { seriousFaults, manoeuvres, vehicleChecks } = data;
+
+    const seriousFaultSumOfSimpleCompetencies = Object.keys(pickBy(seriousFaults)).length;
+    const vehicleCheckSeriousFaults =
+      vehicleChecks ? this.getVehicleChecksFaultCountCatBE(vehicleChecks).seriousFaults : 0;
```
Comment (rossfauldsbjss) [inline src/providers/fault-count/fault-count.ts]: BE not C

File: src/providers/fault-count/fault-count.ts [hunk 5]
```diff
@@ -213,6 +282,24 @@ export class FaultCountProvider {
...
+      check.outcome === CompetencyOutcome.D;
+    }).length : 0;
+
+    const result =
+      dangerousFaultSumOfSimpleCompetencies +
+      this.sumManoeuvreFaults(manoeuvres, CompetencyOutcome.D) +
+      vehicleCheckDangerousFaults;
+
+    return result;
+  }
```
Comment (rossfauldsbjss) [inline src/providers/fault-count/fault-count.ts:295]: think vehicle checks can't have Dangerous faults.

**Review by rossfauldsbjss — DISMISSED** [review]
I think calculations around cat c are wrong. (think you can't have Dangerous and Serious faults for vehicle checks - and that it has similar logic that calculate 1 serious fault if 5 driving faults.  look at Cat B /B+E for how this is done.

File: src/providers/fault-count/fault-count.ts (hunk 4 above)
Comment (amobarak) [reply src/providers/fault-count/fault-count.ts]: Good spot

File: package.json [hunk 6]
```diff
@@ -203,4 +203,4 @@
       "ios"
     ]
   }
-}
+}
```
Comment (TomBillingtonUK) [inline package.json]: why just why

File: src/modules/tests/pass-completion/cat-c/pass-completion.cat-c.reducer.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        ...state,
+        code78Present: true,
+      };
+    case passCompletionActionsCatC.CODE_78_NOT_PRESENT:
+      return {
+        ...state,
+        code78Present: false,
+      };
+    default:
+      return passCompletionReducer(state, action);
```
Comment (TomBillingtonUK) [inline src/modules/tests/pass-completion/cat-c/pass-completion.cat-c.reducer.ts]: duplicate this please , reducers shouldn't be reused

File: src/modules/tests/test-data/cat-be/test-data.cat-be.selector.ts [hunk 8]
```diff
@@ -22,10 +22,10 @@ export const hasManoeuvreBeenCompletedCatBE = (data: CatBEUniqueTypes.TestData)
   );
 };
 
-export const hasEyesightTestBeenCompleted = (data: CatBEUniqueTypes.TestData) => data.eyesightTest.complete;
-
 export const hasEyesightTestGotSeriousFault = (data: CatBEUniqueTypes.TestData) => data.eyesightTest.seriousFault;
 
+export const hasEyesightTestBeenCompleted = (data: CatBEUniqueTypes.TestData) => data.eyesightTest.complete;
```
Comment (TomBillingtonUK) [inline src/modules/tests/test-data/cat-be/test-data.cat-be.selector.ts]: can you revert this change

File: src/pages/non-pass-finalisation/cat-c/non-pass-finalisation.cat-c.page.ts [hunk 9]
```diff
@@ -56,7 +56,7 @@ import { SetTestStatusWriteUp } from '../../../modules/tests/test-status/test-st
 import { SetActivityCode } from '../../../modules/tests/activity-code/activity-code.actions';
 import { BasePageComponent } from '../../../shared/classes/base-page';
 import { getGearboxCategory, isAutomatic, isManual } from
-'../../../modules/tests/vehicle-details/common/vehicle-details.selector';
+  '../../../modules/tests/vehicle-details/common/vehicle-details.selector';
```
Comment (TomBillingtonUK) [inline src/pages/non-pass-finalisation/cat-c/non-pass-finalisation.cat-c.page.ts:63]: Can you change this to 

```
import {
 getGearboxCategory,
 isAutomatic, 
isManual 
} from '../../../modules/tests/vehicle-details/common/vehicle-details.selector';
```

File: src/pages/pass-finalisation/cat-c/pass-finalisation.cat-c.page.html [hunk 10]
```diff
@@ -22,17 +21,18 @@
           </ion-col>
         </ion-row>
 
-        <license-provided [form]="form" (licenseReceived)="provisionalLicenseReceived()"
-          (licenseNotReceived)="provisionalLicenseNotReceived()"> </license-provided>
+        <license-provided [form]="form" [category]="this.category" (licenseReceived)="provisionalLicenseReceived()" (licenseNotReceived)="provisionalLicenseNotReceived()">
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/cat-c/pass-finalisation.cat-c.page.html]: can we move the banner out of the license-provided component and onto the page? We can then avoid the switch statements and complexity in the nested components

File: src/pages/pass-finalisation/components/license-provided/license-provided.ts [hunk 11]
```diff
@@ -1,7 +1,8 @@
 import { Component, OnChanges, Input, Output, EventEmitter } from '@angular/core';
 import { FormGroup, FormControl, Validators } from '@angular/forms';
 import { ProvisionalLicenseReceived, ProvisionalLicenseNotReceived } from
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/components/license-provided/license-provided.ts]: Can we fix the import so it matches the normal multi line imports

File: src/pages/test-report/cat-c/test-report.cat-c.page.html [hunk 12]
```diff
@@ -26,40 +26,25 @@
         </ion-row>
         <ion-row>
           <ion-col>
-            <legal-requirement
```
Comment (TomBillingtonUK) [inline src/pages/test-report/cat-c/test-report.cat-c.page.html]: Can we reformat these so they match what it was before

File: src/providers/test-report-validator/__tests__/test-report-validator.spec.ts [hunk 13]
```diff
@@ -31,7 +31,7 @@ describe('TestReportValidator', () => {
       expect(result).toEqual(false);
     });
     it('should return true if the test report is valid for a Cat BE test', () => {
-      const result = testReportValidatorProvider.isTestReportValid(validTestCatBE , TestCategory.BE);
+      const result = testReportValidatorProvider.isTestReportValid(validTestCatBE, TestCategory.BE);
```
Comment (TomBillingtonUK) [inline src/providers/test-report-validator/__tests__/test-report-validator.spec.ts]: can you revert this change

File: src/pages/pass-finalisation/components/license-provided/__tests__/license-provided.spec.ts [hunk 14]
```diff
@@ -11,19 +13,20 @@ describe('licenseProvidedComponent', () => {
...
-      component = fixture.componentInstance;
-      component.form = new FormGroup({});
-    });
+      .compileComponents()
+      .then(() => {
+        fixture = TestBed.createComponent(LicenseProvidedComponent);
+        component = fixture.componentInstance;
+        component.form = new FormGroup({});
+      });
   }));
```
Comment (appleabs) [inline src/pages/pass-finalisation/components/license-provided/__tests__/license-provided.spec.ts]: Warning banner not used anymore, please remove

**Review by appleabs — DISMISSED** [review]

**Review by appleabs — APPROVED** [review]

File: src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts [hunk 15]
```diff
@@ -174,5 +177,32 @@ describe('PassFinalisationCatCPage', () => {
           .toHaveBeenCalledWith(new PassFinalisationValidationError('notRequiredControl is blank'));
       }));
     });
+
+    describe('showAutomaticBanner', () => {
+      it('should show the automatic banner when it is valid', () => {
+        // component.ionViewDidEnter();
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts]: Comments

File: src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts [hunk 16]
```diff
@@ -174,5 +177,32 @@ describe('PassFinalisationCatCPage', () => {
...
+        component.transmission = TransmissionType.Automatic;
+        component.code78Present = true;
+        expect(component.shouldShowAutomaticBanner()).toEqual(true);
+      });
+    });
+
+    describe('should show manual banner', () => {
+      it('should show the manual banner when it is valid', () => {
+        // component.ionViewDidEnter();
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts]: comments

File: src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts [hunk 17]
```diff
@@ -174,5 +177,32 @@ describe('PassFinalisationCatCPage', () => {
...
+        component.transmission = TransmissionType.Manual;
+        component.code78Present = true;
+        expect(component.shouldShowManualBanner()).toEqual(true);
+      });
+    });
+
+    describe('shouldHideBanner', () => {
+      it('should hide banner when only transmission is selected', () => {
+        // component.ionViewDidEnter();
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/cat-c/_tests_/pass-finalisation.cat-c.page.spec.ts]: comments

File: src/pages/pass-finalisation/cat-c/components/code-78/__tests__/code-78.spec.ts [hunk 18]
```diff
@@ -66,49 +64,5 @@ describe('code78Component', () => {
...
-
-    describe('shouldHideBanner', () => {
-      it('should hide banner when only transmission is selected', () => {
-        component.ngOnChanges();
-        component.transmission = TransmissionType.Manual;
-        fixture.detectChanges();
-        expect(component.shouldHideBanner()).toEqual(true);
-      });
-
-      it('should hide banner when only code78 is selected', () => {
```
Comment (TomBillingtonUK) [inline src/pages/pass-finalisation/cat-c/components/code-78/__tests__/code-78.spec.ts:99]: could you copy this test and the one below to the pass completion page - You deleted 5 tests here but only moved 3 to the other spec file

File: src/pages/test-report/cat-c/test-report.cat-c.page.ts [hunk 19]
```diff
@@ -195,10 +195,10 @@ export class TestReportCatCPage extends BasePageComponent {
         map((data) => {
           // TODO: MES-4287 Use category c function
```
Comment (TomBillingtonUK) [inline src/pages/test-report/cat-c/test-report.cat-c.page.ts]: remove TODO

**Review by TomBillingtonUK — APPROVED** [review]
